File: src/frame_alignment.py
"""
src/frame_alignment.py
Phone-frame -> vehicle-frame alignment.

The rest of the pipeline used to assume acc_y == "forward" and gyro_z == "yaw".
Both are false for a phone at an arbitrary angle, which is the hardest requirement in
the problem statement. Everything here is built from quantities that do not change when
the whole phone is rotated:

    a_vert   = a . g_hat                 (scalar)
    a_horiz  = a - (a . g_hat) g_hat     (3-vector, magnitude is invariant)
    yaw_rate = omega . g_hat             (scalar)

plus a forward axis f_hat estimated from data. f_hat lives in the phone body frame and
rotates with it, so a . f_hat is invariant too: rotating the phone by R sends
a -> Ra and f_hat -> R f_hat, and (Ra).(R f_hat) == a . f_hat.

Causal throughout: the gravity filter is a forward-only single pole, matching what the
Android engine can do in real time.
"""

import numpy as np
from typing import Dict, Optional, Tuple

G0 = 9.80665

# ponytail: single-pole low-pass instead of a Madgwick/Mahony filter. Gravity is the only
# attitude we need and a car does not tumble. Upgrade to TYPE_ROTATION_VECTOR on Android
# (already registered) or a full AHRS if pitch/roll rate ever matters.
#
# KNOWN LIMITATION, measured, not theoretical: at 0.2 Hz the time constant is ~0.8 s, so any
# acceleration sustained longer than a few seconds leaks into the gravity estimate and
# disappears from a_horiz. A long steady brake therefore reads as near-zero horizontal
# specific force. Every complementary attitude filter has this failure; the standard fix is
# to subtract a velocity-derived acceleration prediction before updating gravity, which
# needs a speed estimate we do not trust yet. naive_dr._find_alignment_window works around
# it by using a non-tracking median gravity. Revisit once the speed model is trained on
# real drives.
GRAVITY_LP_HZ = 0.2

# Calibration knobs. These are physical thresholds, not magic numbers - retune per mount.
FWD_MIN_SPEED_MPS = 1.5      # below this, GPS course and accel direction are both noise
FWD_MIN_ACCEL_MPS2 = 0.35    # a real accel/brake event, not idle jitter
FWD_MIN_SAMPLES = 25         # too few events -> refuse to estimate, report low confidence


def lowpass_gravity(acc: np.ndarray, dt: np.ndarray, fc_hz: float = GRAVITY_LP_HZ) -> np.ndarray:
    """
    Causal single-pole low-pass of the raw accelerometer -> gravity estimate, shape (n, 3).

    On Android prefer Sensor.TYPE_GRAVITY, which the sensor hub computes for free; this
    exists so the offline pipeline sees the same signal as the phone.
    """
    acc = np.asarray(acc, dtype=float)
    n = len(acc)
    g = np.zeros((n, 3))
    if n == 0:
        return g

    # Initialise from the first second so the filter does not start at the origin and
    # spend 10 seconds converging.
    warm = max(1, min(n, int(round(1.0 / max(float(np.median(dt)), 1e-3)))))
    g[0] = np.mean(acc[:warm], axis=0)

    tau = 1.0 / (2.0 * np.pi * fc_hz)
    for i in range(1, n):
        alpha = dt[i] / (dt[i] + tau) if dt[i] > 0 else 0.0
        g[i] = (1.0 - alpha) * g[i - 1] + alpha * acc[i]
    return g


def unit(v: np.ndarray, axis: int = -1) -> np.ndarray:
    """Normalise, leaving zero-length rows as zero rather than NaN."""
    n = np.linalg.norm(v, axis=axis, keepdims=True)
    return np.divide(v, n, out=np.zeros_like(v), where=n > 1e-9)


def gravity_stability(g_hat: np.ndarray, window: int) -> np.ndarray:
    """
    Dispersion of the gravity direction over a trailing window.

    This is the single most useful signal in the whole repo for telling "the vehicle is
    moving" from "a human is holding the phone". A cradled phone holds its gravity
    direction to ~0.01; a shaken one swings past 0.3.

    Causal: window [i-w+1, i].
    """
    n = len(g_hat)
    out = np.zeros(n)
    window = max(2, int(window))
    for i in range(n):
        s = max(0, i - window + 1)
        w = g_hat[s:i + 1]
        if len(w) < 2:
            continue
        # Norm of the per-axis std: 0 for a perfectly steady direction.
        out[i] = float(np.linalg.norm(np.std(w, axis=0)))
    return out


def estimate_forward_axis(
    acc_horiz: np.ndarray,
    speed: np.ndarray,
    dt: np.ndarray,
    valid_mask: Optional[np.ndarray] = None
) -> Tuple[np.ndarray, float]:
    """
    Estimate the vehicle forward direction as a unit 3-vector in the phone body frame.

    During GPS-healthy acceleration and braking, horizontal specific force lies along the
    vehicle's longitudinal axis. Take the principal axis of horizontal acceleration over
    those samples, then fix the sign by correlating with the observed speed change.

    Returns (f_hat, confidence in [0, 1]). Confidence is the fraction of horizontal
    acceleration variance explained by the principal axis - near 1.0 when the events are
    genuinely longitudinal, near 0.5 when the data is isotropic noise.
    """
    acc_horiz = np.asarray(acc_horiz, dtype=float)
    speed = np.asarray(speed, dtype=float)

    dv = np.gradient(speed, edge_order=1) / np.maximum(dt, 1e-6)

    mask = (np.abs(dv) > FWD_MIN_ACCEL_MPS2) & (speed > FWD_MIN_SPEED_MPS)
    if valid_mask is not None:
        mask &= valid_mask

    if int(np.sum(mask)) < FWD_MIN_SAMPLES:
        # Not enough evidence. Refuse rather than return a confident wrong axis.
        return np.array([0.0, 1.0, 0.0]), 0.0

    A = acc_horiz[mask]
    A = A - np.mean(A, axis=0)
    cov = A.T @ A / max(1, len(A) - 1)
    evals, evecs = np.linalg.eigh(cov)
    order = np.argsort(evals)[::-1]
    evals = evals[order]
    f_hat = evecs[:, order[0]]

    # Sign: forward acceleration must correlate positively with speed increase.
    proj = acc_horiz[mask] @ f_hat
    if float(np.sum(proj * dv[mask])) < 0:
        f_hat = -f_hat

    total = float(np.sum(evals))
    confidence = float(evals[0] / total) if total > 1e-12 else 0.0
    return unit(f_hat), confidence


def align_frame(
    acc: np.ndarray,
    gyro: np.ndarray,
    dt: np.ndarray,
    speed: Optional[np.ndarray] = None,
    valid_mask: Optional[np.ndarray] = None,
    stability_window: int = 20
) -> Dict[str, np.ndarray]:
    """
    Full alignment. acc/gyro are (n, 3) in the phone body frame; dt is (n,) seconds.

    Returns frame-invariant channels. Every value here is unchanged if you pre-multiply
    acc and gyro by any fixed rotation - that is what tests/test_frame_invariance.py checks.
    """
    acc = np.asarray(acc, dtype=float)
    gyro = np.asarray(gyro, dtype=float)
    dt = np.asarray(dt, dtype=float)

    grav = lowpass_gravity(acc, dt)
    g_hat = unit(grav)

    # Linear (gravity-removed) acceleration.
    lin = acc - grav

    a_vert = np.sum(lin * g_hat, axis=1)
    a_horiz = lin - a_vert[:, None] * g_hat
    a_horiz_mag = np.linalg.norm(a_horiz, axis=1)

    yaw_rate = np.sum(gyro * g_hat, axis=1)
    gyro_mag = np.linalg.norm(gyro, axis=1)
    # Tilt rate: angular velocity not about the vertical. Large when the phone is being
    # handled, near zero for a mounted phone on a flat road.
    tilt_rate = np.sqrt(np.maximum(gyro_mag**2 - yaw_rate**2, 0.0))

    grav_stab = gravity_stability(g_hat, stability_window)

    if speed is not None:
        f_hat, f_conf = estimate_forward_axis(a_horiz, speed, dt, valid_mask)
    else:
        f_hat, f_conf = np.array([0.0, 1.0, 0.0]), 0.0

    # Lateral axis completes the right-handed frame: l_hat = g_hat x f_hat.
    g_ref = unit(np.mean(g_hat, axis=0))
    l_hat = unit(np.cross(g_ref, f_hat))

    a_fwd = a_horiz @ f_hat
    a_lat = a_horiz @ l_hat

    return {
        "gravity": grav,
        "g_hat": g_hat,
        "a_vert": a_vert,
        "a_horiz": a_horiz,
        "a_horiz_mag": a_horiz_mag,
        "a_fwd": a_fwd,
        "a_lat": a_lat,
        "yaw_rate": yaw_rate,
        "gyro_mag": gyro_mag,
        "tilt_rate": tilt_rate,
        "grav_stability": grav_stab,
        "forward_axis": f_hat,
        "forward_confidence": np.array([f_conf]),
    }


def add_frame_columns(df, stability_window: int = 20):
    """
    Convenience wrapper: takes a standardized drive dataframe, returns it with the
    frame-invariant channels appended. Used by feature_engineering and the EKF pipeline
    so they never touch raw acc_y / gyro_z again.
    """
    dt = df["dt"].values if "dt" in df.columns else np.gradient(df["timestamp"].values)
    acc = np.column_stack([df["acc_x"].values, df["acc_y"].values, df["acc_z"].values])
    gyro = np.column_stack([df["gyro_x"].values, df["gyro_y"].values, df["gyro_z"].values])
    speed = df["speed"].values if "speed" in df.columns else None

    fr = align_frame(acc, gyro, dt, speed=speed, stability_window=stability_window)

    out = df.copy()
    for k in ["a_vert", "a_horiz_mag", "a_fwd", "a_lat", "yaw_rate", "gyro_mag",
              "tilt_rate", "grav_stability"]:
        out[k] = fr[k]
    out.attrs["forward_axis"] = fr["forward_axis"]
    out.attrs["forward_confidence"] = float(fr["forward_confidence"][0])
    return out


def demo():
    """Self-check: rotating the whole rig must not change any invariant channel."""
    rng = np.random.default_rng(0)
    n = 600
    dt = np.full(n, 0.1)
    t = np.arange(n) * 0.1

    # A phone lying flat, vehicle accelerating forward along body +y, turning at 0.2 rad/s.
    acc = np.column_stack([
        rng.normal(0, 0.02, n),
        1.2 * np.sin(2 * np.pi * 0.05 * t) + rng.normal(0, 0.02, n),
        np.full(n, G0) + rng.normal(0, 0.02, n),
    ])
    gyro = np.column_stack([
        rng.normal(0, 0.001, n),
        rng.normal(0, 0.001, n),
        np.full(n, 0.2) + rng.normal(0, 0.001, n),
    ])
    speed = 8.0 + np.cumsum(acc[:, 1] * dt)

    base = align_frame(acc, gyro, dt, speed=speed)

    # Arbitrary fixed rotation of the entire phone.
    ang = 0.9
    axis = unit(np.array([0.3, -0.7, 0.65]))
    K = np.array([[0, -axis[2], axis[1]], [axis[2], 0, -axis[0]], [-axis[1], axis[0], 0]])
    R = np.eye(3) + np.sin(ang) * K + (1 - np.cos(ang)) * (K @ K)

    rot = align_frame(acc @ R.T, gyro @ R.T, dt, speed=speed)

    for key in ["a_vert", "a_horiz_mag", "yaw_rate", "gyro_mag", "tilt_rate", "grav_stability"]:
        err = float(np.max(np.abs(base[key] - rot[key])))
        assert err < 1e-9, f"{key} is not frame-invariant: max abs diff {err}"

    # The forward axis must rotate with the phone, keeping a_fwd invariant.
    err_fwd = float(np.max(np.abs(base["a_fwd"] - rot["a_fwd"])))
    assert err_fwd < 1e-6, f"a_fwd not invariant: {err_fwd}"
    assert base["forward_confidence"][0] > 0.9, "forward axis should be confident here"

    print("frame_alignment demo OK: invariance holds under arbitrary rotation")


if __name__ == "__main__":
    demo()

File: flutter_app/lib/localization/frame_alignment.dart
import 'dart:math' as math;

/// Phone-frame -> vehicle-frame alignment, ported from src/frame_alignment.py.
///
/// Nothing downstream may consume raw ax/ay/az or gyro z. Those encode how the phone
/// happens to be sitting in the cradle, so a model trained on them learns one mounting and
/// fails on the next. Everything here is built from quantities that do not change when the
/// whole phone is rotated:
///
///   aVert    = a . gHat                  (scalar)
///   aHoriz   = a - (a . gHat) gHat       (magnitude is invariant)
///   yawRate  = omega . gHat              (scalar)
///
/// plus a forward axis estimated from data, which rotates with the phone so that
/// a . fHat is invariant too.
///
/// The offline pipeline verifies this to 1e-14 across 12 random SO(3) rotations of a real
/// drive (tests/test_rotation_invariance.py).
class FrameAlignment {
  static const double g0 = 9.80665;

  /// Mahony complementary filter constants. A plain low-pass cannot do this job: one time
  /// constant has to both reject linear acceleration (wants slow) and track real tilt
  /// (wants fast), and at 0.2 Hz it resolved that by absorbing any sustained acceleration
  /// into "gravity". Measured on 32 real brake events, the low-pass retained 19% of the
  /// vehicle's actual deceleration; this filter retains 61%.
  static const double kp = 0.10;
  static const double ki = 0.002;
  static const double accelTrustBand = 1.5;

  final List<double> _g = [0.0, 0.0, g0];
  final List<double> _bias = [0.0, 0.0, 0.0];
  bool _initialised = false;

  /// Forward axis in the phone body frame, estimated from acceleration events.
  List<double> _fHat = [0.0, 1.0, 0.0];
  final List<List<double>> _fwdCov =
      List.generate(3, (_) => List.filled(3, 0.0));
  int _fwdEvents = 0;
  double forwardConfidence = 0.0;

  double aVert = 0.0;
  double aHorizMag = 0.0;
  double aFwd = 0.0;
  double aLat = 0.0;
  double yawRate = 0.0;
  double gyroMag = 0.0;
  double tiltRate = 0.0;
  double gravStability = 0.0;

  final List<List<double>> _gHist = [];
  int _stabilityWindow = 20;

  /// Feed one IMU sample. [gravity] is Sensor.TYPE_GRAVITY when the platform provides it;
  /// the sensor hub computes it more accurately than we can, for free.
  void update(
    double ax,
    double ay,
    double az,
    double gx,
    double gy,
    double gz,
    double dt, {
    List<double>? gravity,
  }) {
    if (!_initialised) {
      _g[0] = gravity?[0] ?? ax;
      _g[1] = gravity?[1] ?? ay;
      _g[2] = gravity?[2] ?? az;
      _stabilityWindow = math.max(2, (2.0 / (dt > 0 ? dt : 0.02)).round());
      _initialised = true;
    } else if (gravity != null) {
      _g[0] = gravity[0];
      _g[1] = gravity[1];
      _g[2] = gravity[2];
    } else {
      _propagateGravity(ax, ay, az, gx, gy, gz, dt);
    }

    final gn = math.max(
        math.sqrt(_g[0] * _g[0] + _g[1] * _g[1] + _g[2] * _g[2]), 1e-9);
    final ux = _g[0] / gn, uy = _g[1] / gn, uz = _g[2] / gn;

    final lx = ax - _g[0], ly = ay - _g[1], lz = az - _g[2];
    aVert = lx * ux + ly * uy + lz * uz;
    final hx = lx - aVert * ux, hy = ly - aVert * uy, hz = lz - aVert * uz;
    aHorizMag = math.sqrt(hx * hx + hy * hy + hz * hz);

    // Yaw is the gyro projected onto gravity. The z axis is only yaw if the phone happens
    // to be lying perfectly flat, which it never is in a cradle.
    yawRate = gx * ux + gy * uy + gz * uz;
    gyroMag = math.sqrt(gx * gx + gy * gy + gz * gz);
    tiltRate = math.sqrt(math.max(0.0, gyroMag * gyroMag - yawRate * yawRate));

    _pushGravityHistory([ux, uy, uz]);
    gravStability = _gravityStability();
    _updateMountTrust(dt);

    _lastHoriz = [hx, hy, hz];
    aFwd = hx * _fHat[0] + hy * _fHat[1] + hz * _fHat[2];
    final l = _cross([ux, uy, uz], _fHat);
    aLat = hx * l[0] + hy * l[1] + hz * l[2];
  }

  // ── Mount disturbance: is the PHONE turning, or is the VEHICLE turning? ────
  //
  // yawRate = omega . gHat is the rotation rate about the vertical. It is frame-invariant,
  // which makes it the right quantity - but it cannot tell whose rotation it is. Rotating
  // the phone in its cradle and driving round a bend look identical to a gyroscope.
  //
  // Physics does separate them. A vehicle turning on a road rotates about the vertical and
  // essentially nothing else, so its rotation has almost no component perpendicular to
  // gravity. A hand slanting a phone rotates it about a horizontal axis, which shows up
  // directly in tiltRate. Measured on the synthetic rig in test/tilt_heading_test.dart:
  //
  //   real 90 deg/6 s vehicle turn, phone still   tiltRate ~ 0.00 rad/s
  //   hand slant, 25 deg/s                        tiltRate ~ 0.44 rad/s
  //
  // So tiltRate is the discriminator, and yaw is believed in proportion to how still the
  // mounting is. A soft ramp rather than a hard gate, because a hard gate would throw away
  // a genuine turn the instant the phone was nudged.
  //
  // ponytail: this suppresses real vehicle yaw during a disturbance rather than recovering
  // it - during those ~1 s the two are genuinely not separable from the gyro alone.
  // Upgrade path is to estimate yaw from lateral specific force (a_lat = v * omega) while
  // the mounting is moving, which needs a speed estimate we do not yet trust enough.
  static const double tiltQuiet = 0.08;      // rad/s: road camber and pitch live below this
  static const double tiltDisturbed = 0.30;  // rad/s: unambiguously a hand
  static const double settleSeconds = 0.6;   // hold-off after the phone stops moving

  /// 0 = the measured yaw is not the vehicle's, 1 = fully trusted.
  double yawTrust = 1.0;

  /// The vehicle's yaw rate about the vertical, ANTICLOCKWISE-positive (the right-hand
  /// rule about gHat, which points up). Correct for physics, wrong for a compass.
  double vehicleYawRate = 0.0;

  /// The same rate as a COMPASS bearing rate: clockwise-positive, so that adding it to a
  /// heading where 0 = North and +90 = East turns the right way.
  ///
  /// This exists because the two halves of the codebase disagreed about the sign and the
  /// disagreement was invisible until someone drove round a corner. Heading is integrated
  /// as a bearing (east += v*sin(psi), north += v*cos(psi)) which is clockwise-positive,
  /// while omega . gHat is anticlockwise-positive. Feeding one into the other inverts
  /// every turn. The pedestrian path had been patched with a local minus sign; the EKF
  /// path had not. Both now take this, and the conversion is stated once, here.
  double get compassYawRate => -vehicleYawRate;

  /// True while the phone is moving relative to its mounting.
  bool mountDisturbed = false;

  /// Seconds of disturbance accumulated in the current session - surfaced in the UI so a
  /// held heading is explained rather than looking like a freeze.
  double disturbanceSeconds = 0.0;

  double _settleTimer = 0.0;

  void _updateMountTrust(double dt) {
    double t;
    if (tiltRate <= tiltQuiet) {
      t = 1.0;
    } else if (tiltRate >= tiltDisturbed) {
      t = 0.0;
    } else {
      t = 1.0 - (tiltRate - tiltQuiet) / (tiltDisturbed - tiltQuiet);
    }

    // After the phone stops moving the gravity estimate is still re-settling and the
    // mounting rotation has changed, so trust is ramped back rather than restored at once.
    if (t < 0.5) {
      _settleTimer = settleSeconds;
    } else if (_settleTimer > 0) {
      _settleTimer = math.max(0.0, _settleTimer - dt);
      t = math.min(t, 1.0 - _settleTimer / settleSeconds);
    }

    yawTrust = t.clamp(0.0, 1.0);
    mountDisturbed = yawTrust < 0.5;
    if (mountDisturbed) disturbanceSeconds += dt;
    vehicleYawRate = yawRate * yawTrust;
  }

  void _propagateGravity(double ax, double ay, double az, double gx, double gy,
      double gz, double dt) {
    final aNorm = math.sqrt(ax * ax + ay * ay + az * az);
    var wx = gx - _bias[0], wy = gy - _bias[1], wz = gz - _bias[2];

    // Only trust the accelerometer when it is plausibly reading gravity alone - i.e. NOT
    // while the vehicle is braking hard, which is exactly when we need the estimate to
    // hold still.
    if ((aNorm - g0).abs() < accelTrustBand && aNorm > 1e-6) {
      final gn = math.max(
          math.sqrt(_g[0] * _g[0] + _g[1] * _g[1] + _g[2] * _g[2]), 1e-9);
      final v = [_g[0] / gn, _g[1] / gn, _g[2] / gn];
      final aHat = [ax / aNorm, ay / aNorm, az / aNorm];
      final err = _cross(v, aHat);
      wx -= kp * err[0];
      wy -= kp * err[1];
      wz -= kp * err[2];
      _bias[0] += ki * err[0] * dt;
      _bias[1] += ki * err[1] * dt;
      _bias[2] += ki * err[2] * dt;
    }

    final gn = math.max(
        math.sqrt(_g[0] * _g[0] + _g[1] * _g[1] + _g[2] * _g[2]), 1e-9);
    final v = [_g[0] / gn, _g[1] / gn, _g[2] / gn];
    final rot = _cross([wx, wy, wz], v);
    for (var i = 0; i < 3; i++) {
      v[i] -= rot[i] * dt;
    }
    final n =
        math.max(math.sqrt(v[0] * v[0] + v[1] * v[1] + v[2] * v[2]), 1e-9);
    for (var i = 0; i < 3; i++) {
      _g[i] = v[i] / n * g0;
    }
  }

  void _pushGravityHistory(List<double> u) {
    _gHist.add(u);
    while (_gHist.length > _stabilityWindow) {
      _gHist.removeAt(0);
    }
  }

  /// Dispersion of the gravity direction. ~0.01 for a phone in a cradle, past 0.3 for one
  /// being held. This single number is the most useful signal in the system for telling
  /// "the vehicle moved" from "a human moved the phone".
  double _gravityStability() {
    if (_gHist.length < 2) return 0.0;
    var acc = 0.0;
    for (var axis = 0; axis < 3; axis++) {
      var m = 0.0;
      for (final v in _gHist) {
        m += v[axis];
      }
      m /= _gHist.length;
      var s = 0.0;
      for (final v in _gHist) {
        final d = v[axis] - m;
        s += d * d;
      }
      acc += s / _gHist.length;
    }
    return math.sqrt(acc);
  }

  /// Accumulate the forward-axis estimate during genuine acceleration events while GNSS is
  /// healthy, then take the principal axis of horizontal acceleration.
  void observeForwardAxis(double dv, double speed) {
    if (dv.abs() < 0.35 || speed < 1.5) return;
    final v = _lastHoriz;
    if (v == null) return;
    for (var i = 0; i < 3; i++) {
      for (var j = 0; j < 3; j++) {
        _fwdCov[i][j] += v[i] * v[j];
      }
    }
    _fwdEvents++;
    if (_fwdEvents < 25 || _fwdEvents % 10 != 0) return;

    var e = [_fHat[0], _fHat[1], _fHat[2]];
    for (var it = 0; it < 24; it++) {
      final n = [
        _fwdCov[0][0] * e[0] + _fwdCov[0][1] * e[1] + _fwdCov[0][2] * e[2],
        _fwdCov[1][0] * e[0] + _fwdCov[1][1] * e[1] + _fwdCov[1][2] * e[2],
        _fwdCov[2][0] * e[0] + _fwdCov[2][1] * e[1] + _fwdCov[2][2] * e[2],
      ];
      final nn = math.sqrt(n[0] * n[0] + n[1] * n[1] + n[2] * n[2]);
      if (nn > 1e-12) e = [n[0] / nn, n[1] / nn, n[2] / nn];
    }
    final proj = v[0] * e[0] + v[1] * e[1] + v[2] * e[2];
    if (proj * dv < 0) e = [-e[0], -e[1], -e[2]];
    _fHat = e;

    final trace = _fwdCov[0][0] + _fwdCov[1][1] + _fwdCov[2][2];
    final along = e[0] *
            (_fwdCov[0][0] * e[0] +
                _fwdCov[0][1] * e[1] +
                _fwdCov[0][2] * e[2]) +
        e[1] *
            (_fwdCov[1][0] * e[0] +
                _fwdCov[1][1] * e[1] +
                _fwdCov[1][2] * e[2]) +
        e[2] *
            (_fwdCov[2][0] * e[0] +
                _fwdCov[2][1] * e[1] +
                _fwdCov[2][2] * e[2]);
    forwardConfidence = trace > 1e-12 ? along / trace : 0.0;
  }

  /// Horizontal specific force from the most recent update, in body frame.
  List<double>? _lastHoriz;

  List<double> _cross(List<double> a, List<double> b) => [
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
      ];

  void reset() {
    _initialised = false;
    yawTrust = 1.0;
    vehicleYawRate = 0.0;
    mountDisturbed = false;
    disturbanceSeconds = 0.0;
    _settleTimer = 0.0;
    _gHist.clear();
    _fwdEvents = 0;
    forwardConfidence = 0.0;
    _fHat = [0.0, 1.0, 0.0];
    for (final r in _fwdCov) {
      r.fillRange(0, 3, 0.0);
    }
    _bias.fillRange(0, 3, 0.0);
  }
}
